### PR TITLE
Improves object fastening logic and directional window code.

### PR DESCRIPTION
**Pull request by Timberpoes**
# I CODED THIS WITH MAXIMUM SMOOTHBRAIN, I HAVE NO CLUE WHAT I'M DOING, SEND HELP.

## About The Pull Request

Reverts #56205
Replaces the behaviour in this PR with a more generic solution.

Allows you to secure objects and stuff on the same tile as dense things by adding a bonus check to is_blocked_turf where if you're specifying a source_atom, it'll also check if that atom CanPass any dense atoms on the turf. This allows you to wrench machines to turfs with directional windoors, chargers to tables, etc. in a generic way. If that atom can legitimately pass all dense objects on that turf, it can be anchored to it.

You can now push non-anchored directional windows where their dir would otherwise have blocked you from doing this before.

You can now drag and push non-anchored directional windows onto turfs with other anchored and non-anchored directional windows and their equivalents.

Cleans up a bit of code here and there. Docs a few things I've touched as well.

## Why It's Good For The Game

You can now secure things ontop of dense things that they can legitimately pass in a generic manner. This is better than snowflaking constant type checks whenever we do or do not want something non-dense anchored. Probably. Bet it has a ton of unintended consequences I didn't consider.
YLGznuKFtK

As a result, you can also now secure machines ontop of turfs with directional windows. Directional windows are now a lot less finnicky in general and will not block turfs when the atom could pass onto that turf.
ykIYQpN8wW

You can now properly arrange directional windows. You can push them. You can pull them. You can drag them onto the same tile where they're compatible with other directional windows on that tile.
mUaupwFmQt

## Changelog
:cl:
fix: You can now push unanchored directional windows in scenarios where they would previously block your movement.
tweak: You can now anchor machines on the same tiles as directional windows where previously this was impossible.
tweak: You can fasten non-dense objects (such as grinders and rechargers) onto tiles with dense things (such as tables) more reliably than before.
/:cl:

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
Does this prevent/reintroduce #49405?

File: code/game/objects/structures/windoor_assembly.dm [hunk 1]
```diff
@@ -44,27 +43,28 @@
...
 		return FALSE
 
+	if(istype(mover, /obj/structure/window))
+		var/obj/structure/window/moved_window = mover
+		return valid_window_location(loc, moved_window.dir, is_fulltile = moved_window.fulltile)
+
+	if(istype(mover, /obj/structure/windoor_assembly) || istype(mover, /obj/machinery/door/window))
+		return valid_window_location(loc, mover.dir, is_fulltile = FALSE)
+
+	return TRUE
```
Comment (LemonInTheDark) [inline code/game/objects/structures/windoor_assembly.dm:66]: Doesn't this mean that if the parent returns false, but none of the cases are true it returns true by default? Why return TRUE at the end?

File: code/game/objects/structures/windoor_assembly.dm [hunk 2]
```diff
@@ -44,27 +43,28 @@
...
-		if(!valid_window_location(loc, W.ini_dir))
-			return FALSE
-	else if(istype(mover, /obj/structure/windoor_assembly))
-		var/obj/structure/windoor_assembly/W = mover
-		if(!valid_window_location(loc, W.ini_dir))
-			return FALSE
-	else if(istype(mover, /obj/machinery/door/window) && !valid_window_location(loc, mover.dir))
+
+	if(get_dir(loc, target) == dir)
 		return FALSE
```
Comment (LemonInTheDark) [inline code/game/objects/structures/windoor_assembly.dm:57]: This seems like an inappropriate conversion, if we're facing the same direction as the mover shouldn't we let the parent call decide?

File: code/game/objects/structures/window.dm [hunk 3]
```diff
@@ -99,33 +97,25 @@
...
-			return FALSE
-	else if(istype(mover, /obj/machinery/door/window) && !valid_window_location(loc, mover.dir))
+
+	if(fulltile)
 		return FALSE
-	else if(attempted_dir != dir)
-		return TRUE
+
+	if(get_dir(loc, target) == dir)
+		return FALSE
```
Comment (LemonInTheDark) [inline code/game/objects/structures/window.dm:109]: Same question here as before

File: code/modules/mob/living/living.dm [hunk 4]
```diff
@@ -205,30 +205,32 @@
 	if(!client && (mob_size < MOB_SIZE_SMALL))
 		return
 	now_pushing = TRUE
-	var/t = get_dir(src, AM)
+	var/dir_to_target = get_dir(src, AM)
+	if(!dir_to_target)
+		dir_to_target = dir
```
Comment (LemonInTheDark) [inline code/modules/mob/living/living.dm:210]: What happen here?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Mostly questions about the logic of your checks and how returns interact with it, looking good otherwise

**Comment by Timberpoes** [comment]
> 
> 
> Does this prevent/reintroduce #49405?

> 
> 
> Does this prevent/reintroduce #49405?

In my current set of testing, no it does not. You can't make frames stacked ontop of eachother and frames are dense with nothing able to pass through them, so you can't drag a frame over something non-dense and wrench both.

However... Let's just say I've found a way to replicate that behaviour on live.

**Comment by RaveRadbury** [comment]
Will this allow vend-a-trays to be moved onto tables? Paystands? Grinders? Microwaves? There are a number of machines that require you to build a table underneath them, that don't *quite* make sense.

File: code/modules/unit_tests/rcd.dm [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	var/turf/adjacent_turf = adjacent_turfs[1]
+
+	for(var/i in 1 to 10)
+		adjacent_turf.rcd_act(engineer, rcd, rcd.mode)
+
+	var/frame_count = 0
+	for(var/obj/structure/frame/machine_frame in adjacent_turf.contents)
+		frame_count++
+
+	TEST_ASSERT(frame_count == 1, "Expected RCD test to end up with 1 machine frame. It instead created [frame_count] machine frames.")
```
Comment (Jared-Fogle) [inline code/modules/unit_tests/rcd.dm:23]: The macro will show the values you are looking for.

```suggestion
	TEST_ASSERT_EQUAL(frame_count, 1, "One machine frame was not created.")
```

File: code/modules/unit_tests/rcd.dm [hunk 6]
```diff
@@ -0,0 +1,23 @@
+/datum/unit_test/rcd/Run()
```
Comment (Jared-Fogle) [inline code/modules/unit_tests/rcd.dm:1]: Can you document what this is trying to test? It's not obvious through the code to me.

**Review by Mothblocks — DISMISSED** [review]
Reviewing unit tests.

File: code/game/objects/structures/window.dm (hunk 3 above)
Comment (Timberpoes) [reply code/game/objects/structures/window.dm:109]: In this one, if(.) return was part of the original proc. The only possible return value for `return` at this point was FALSE as . has already returned on TRUE.

File: code/game/objects/structures/windoor_assembly.dm (hunk 2 above)
Comment (Timberpoes) [reply code/game/objects/structures/windoor_assembly.dm:57]: Good catch, copypasta prevails. I've tweaked this to match the original flow of logic.

File: code/game/objects/structures/windoor_assembly.dm (hunk 1 above)
Comment (Timberpoes) [reply code/game/objects/structures/windoor_assembly.dm:66]: Good catch, tweaked.

File: code/modules/mob/living/living.dm (hunk 4 above)
Comment (Timberpoes) [reply code/modules/mob/living/living.dm:210]: I've added a comment to explain the scenario where this happens. Feel free to check it again in light of that.

**Comment by Timberpoes** [comment]
Unit tests removed, will be incorporated in another PR.

**Comment by Timberpoes** [comment]
> 
> 
> Will this allow vend-a-trays to be moved onto tables? Paystands? Grinders? Microwaves? There are a number of machines that require you to build a table underneath them, that don't _quite_ make sense.

This allows Microwaves, Grinders, Cell and Weapon Chargers to be put on tables with this generic logic tweak. It does not allow paystands to be put on tables or vend-a-trays - So they can't be secured to tables. It does not allow certain non-dense machines like Holopads to be put on tables. The object has to be able to pass onto a turf when all of the turf's contents are considered, to be fastened to it.

File: code/modules/unit_tests/rcd.dm (hunk 6 above)
Comment (Timberpoes) [reply code/modules/unit_tests/rcd.dm]: Unit test moved to https://github.com/tgstation/tgstation/pull/56287

File: code/modules/unit_tests/rcd.dm (hunk 5 above)
Comment (Timberpoes) [reply code/modules/unit_tests/rcd.dm]: Unit test moved to https://github.com/tgstation/tgstation/pull/56287

File: code/game/atoms_movable.dm [hunk 7]
```diff
@@ -69,6 +69,9 @@
 	///Highest-intensity light affecting us, which determines our visibility.
 	var/affecting_dynamic_lumi = 0
 
+	/// Whether this atom should have its dir automatically changed when it moves. Setting this to FALSE allows for things such as directional windows to retain dir on moving without snowflake code all of the place.
+	var/set_dir_on_move = TRUE
```
Comment (Qustinnus) [inline code/game/atoms_movable.dm:73]: minor point that i guess might be out of scope but I feel like movement related bools should really just be condensed into a bitfield to save memory

**Comment by coiax** [comment]
Possibly out of scope, but some of these "user stories" would be excellent candidates for testing. For example:

image

Should be able to set up a scenario where the ORM is pushed and then wrenched, and it should be fastenable?

**Comment by Timberpoes** [comment]
@LemonInTheDark If you get a moment, please swing by and re-check this in light of my changes/comments on your review requests

**Comment by LemonInTheDark** [comment]
:+

**Comment by Timberpoes** [comment]
Merged as part of #56287

**Pull request closed without merging**